nexus/db-queries/benches/sled_reservation.rs

```diff
@@ -0,0 +1,529 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Benchmarks creating sled reservations
+
+use criterion::black_box;
+use criterion::{criterion_group, criterion_main, Criterion, SamplingMode};
+use nexus_db_queries::authz;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use omicron_common::api::external;
+use omicron_test_utils::dev;
+use omicron_uuid_kinds::InstanceUuid;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Duration;
+use std::time::Instant;
+use tokio::sync::Barrier;
+
+mod harness;
+
+use harness::db_utils::create_affinity_group;
+use harness::db_utils::create_affinity_group_member;
+use harness::db_utils::create_anti_affinity_group;
+use harness::db_utils::create_anti_affinity_group_member;
+use harness::db_utils::create_instance_record;
+use harness::db_utils::create_reservation;
+use harness::db_utils::delete_affinity_group;
+use harness::db_utils::delete_anti_affinity_group;
+use harness::db_utils::delete_instance_record;
+use harness::db_utils::delete_reservation;
+use harness::db_utils::max_resource_request_count;
+use harness::TestHarness;
+
+/////////////////////////////////////////////////////////////////
+//
+// PARAMETERS
+//
+// Describes varations between otherwise shared test logic
+
+#[derive(Clone, Hash, Eq, PartialEq)]
+enum GroupType {
+    Affinity,
+    AntiAffinity,
+}
+
+#[derive(Clone, Hash, Eq, PartialEq)]
+struct GroupInfo {
+    // Name of affinity/anti-affinity group
+    name: &'static str,
+    // Policy of the group
+    policy: external::AffinityPolicy,
+    // Type of group
+    flavor: GroupType,
+}
+
+#[derive(Clone)]
+struct InstanceGroups {
+    // An instance should belong to all these groups.
+    belongs_to: Vec<GroupInfo>,
+}
+
+#[derive(Clone)]
+struct InstanceGroupPattern {
+    description: &'static str,
+    // These "instance group settings" should be applied.
+    //
+    // This is called "stripe" because we rotate through these groups.
+    //
+    // For example, with a stripe.len() == 2...
+    //   - Instance 0 belongs to the groups in stripe[0]
+    //   - Instance 1 belongs to the groups in stripe[1]
+    //   - Instance 2 belongs to the groups in stripe[0]
+    //   - etc
+    stripe: Vec<InstanceGroups>,
+}
+
+#[derive(Clone)]
+struct TestParams {
+    // Number of vmms to provision from the task-under-test
+    vmms: usize,
```

Comment:
Just to make sure I'm following everything: this also determines the number of *instances* the test creates, and each task will end up allocating one VMM for each of these instances (i.e., the VMMs/instances are not partitioned by task). Is that correct? (I'm 99% sure it is from the phrase "task-under-test," just making sure I haven't missed something.)

Reply:
That is correct, each "task" operates on a totally distinct set of instances/vmms from the other tasks.